10-Profiling.md

```diff
@@ -137,129 +141,182 @@ If you just want to see how long it takes your program to run, this is fine.
+- Two functions that repeatedly compute the $20^{th}$ Fibonnaci number (`do_iterative` and `do_recursive`)
+- `main`, which calls `fib_iterative` 200,000 times and `fib_recursive` 200 times
 
 ~~~{.cpp .numberLines}
 #include <iostream>
 using namespace std;
 
-void new_func1()
+long fib_iterative(const long n)
 {
-    cout << "Inside new_func1" << endl;
-    for (int i = 0; i < 2000000000; i++)
-    {
-    }
-    return;
+  long fib = 1;
+  long prev_fib = 0;
+
+  for(long i = 0; i < n; i++)
+  {
+    long tmp = fib;
+    fib += prev_fib;
+    prev_fib = tmp;
+  }
+
+  return fib;
 }
 
-void func1()
+void do_iterative(const int count)
 {
-    cout << "Inside func1" << endl;
-    for (int i = 0; i < 2000000000; i++)
-    {
-    }
-    new_func1();
-    return;
+  volatile long fib;
+  cout << "Iterative: ";
+  for(int i = 0; i < count; i++)
+    fib = fib_iterative(20);
+  cout << fib << endl;
 }
 
-void func2()
+long fib_recursive(const long n)
 {
-    cout << "Inside func2" << endl;
-    for (int i = 0; i < 2000000000; i++)
-    {
-    }
-    return;
+  if(n <= 1) return 1;
+  return fib_recursive(n-1) + fib_recursive(n-2);
 }
 
-int main(void)
+void do_recursive(const int count)
 {
-    cout << "Inside main" << endl;
-    for (int i = 0; i < 2000000000; i++)
-    {
-    }
-    func1();
-    func2();
-    return 0;
+  volatile long fib;
+  cout << "Recursive: ";
+  for(int i = 0; i < count; i++)
+    fib = fib_recursive(20);
+  cout << fib << endl;
+}
+
+int main()
+{
+  const int count = 200;
+
+  do_iterative(count * 1000);
+
+  do_recursive(count);
+  return 0;
 }
 ~~~
 
-In order to use `gprof` we have to pass an additional flag to `g++`.
-The `-pg` flag tells `g++` to record profile information whenever our compiled program runs.
-Each time you run your program it will generate a file called `gmon.out` which contains
-information that can then be interpreted by the `gprof` command.
+In order to use `gperftool`, we have to link its library with our code using the `-lprofiler` flag.
+We'll also want to use the `-g` flag so that `kcachegrind` can show us our source code when we're viewing our profiling output!
 
 Let's compile the program above.
-We'll assume it's called `main.cpp`.
+We'll assume it's called `fibonnaci.cpp`.
 
 ~~~shell
 $ ls
-main.cpp
-$ g++ -pg -o main main.cpp
+fibonnaci.cpp
+$ g++ -g -lprofiler -o  fibonnaci.cpp
 $ ls
-main    main.cpp
+fibonnaci    fibonnaci.cpp
 ~~~
 
-In order to generate `gmon.out`, we need to run `main`.
+By default, `gperftool` doesn't do anything when you execute your program.
+To get it to profile your code, you will need to set an environment variable, `CPUPROFILE`, to the name of the file where you want to store your profile output.
+We'll store ours in a file named `gperftool.prof`:
 
 ~~~shell
+$ CPUPROFILE=gperftool.prof ./fibonnaci
+Iterative: 10946
+Recursive: 10946
+PROFILE: interrupts/evictions/bytes = 4/0/608
```

Comment:
Should everyone also see `4/0/608` or is this system dependent?

Reply:
Probably system dependent?